File: interviews/lc/10-valid-sudoku.cpp
#include <complex>
#include <cstdlib>
#include "lc-utility.h"

inline bool isValidNum(char c) {
  return c != '.';
}

inline int char2num(char c) {
  return c - '0';
}

inline int getBox(int row, int col) {
  int box = (row / 3) * 3;
  if (col < 3) {
    ;
  } else if (col < 6) {
    box += 1;
  } else if (col < 9) {
    box += 2;
  }
  return box;
}

bool isValidSudoku(const vector<vector<char>>& board) {
  int rows[9];
  int columns[9];
  int boxes[9];
  memset(rows, 0, sizeof(int) * 9);
  memset(columns, 0, sizeof(int) * 9);
  memset(boxes, 0, sizeof(int) * 9);

  for (int row = 0; row < 9; row++) {
    for (int col = 0; col < 9; col++) {
      const char c = board[row][col];
      if (!isValidNum(c)) {
        continue;
      }
      const int num = char2num(c);
      const int mask = 1 << (num - 1);

      if ((rows[row] & mask) > 0) {
        return false;
      }
      if ((columns[col] & mask) > 0) {
        return false;
      }
      const auto box = getBox(row, col);
      if ((boxes[box] & mask) > 0) {
        return false;
      }

      rows[row] |= mask;
      columns[col] |= mask;
      boxes[box] |= mask;
    }
  }
  return true;
}

void printBoard(const vector<vector<char>>& board) {
  for (int row = 0; row < board.size(); row++) {
    for (int col = 0; col < board[row].size(); col++) {
      const char c = board[row][col];
      cout << c << " ";
    }
    cout << endl;
  }
  cout << endl;
}

int main(int argc, char* argv[]) {
  vector<vector<char>> board = {{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                                {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                                {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                                {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                                {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                                {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                                {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                                {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                                {'.', '.', '.', '.', '8', '.', '.', '7', '9'}};
  cout << "Board: " << endl;
  printBoard(board);
  cout << "IsValid: " << (isValidSudoku(board) ? "true" : "false") << endl;

  return EXIT_SUCCESS;
}

File: interviews/lc/10-valid-sudoku.py

def isValidNumber(c: str) -> bool:
    return c != '.'


def char2num(c: str) -> int:
    return ord(c[0]) - ord('0')


def getBox(row: int, col: int) -> int:
    box = int(row / 3) * 3
    if col >= 3 and col < 6:
        box += 1
    elif col >= 6 and col < 9:
        box += 2
    return box
    

def isValidSudoku(board: list[list[str]]) -> bool:
    rows = [0 for _ in range(9)]
    cols = [0 for _ in range(9)]
    boxes = [0 for _ in range(9)]

    for row in range(9):
        for col in range(9):

            c = board[row][col]
            if not isValidNumber(c):
                continue

            num = char2num(c)
            mask = 1 << (num - 1)
            
            if rows[row] & mask:
                return False
            if cols[col] & mask:
                return False
            box = getBox(row, col)
            if boxes[box] & mask:
                return False

            rows[row] |= mask
            cols[col] |= mask
            boxes[box] |= mask

    return True


def runTestCase(board: list[list[str]]) -> None:
    for row in board:
        print(" ".join([str(n) for n in row]))
    isValid = isValidSudoku(board)
    print(f"IsValid: {isValid}")
        

if __name__ == "__main__":
    runTestCase([['5', '3', '.', '.', '7', '.', '.', '.', '.'],
                ['6', '.', '.', '1', '9', '5', '.', '.', '.'],
                ['.', '9', '8', '.', '.', '.', '.', '6', '.'],
                ['8', '.', '.', '.', '6', '.', '.', '.', '3'],
                ['4', '.', '.', '8', '.', '3', '.', '.', '1'],
                ['7', '.', '.', '.', '2', '.', '.', '.', '6'],
                ['.', '6', '.', '.', '.', '.', '2', '8', '.'],
                ['.', '.', '.', '4', '1', '9', '.', '.', '5'],
                ['.', '.', '.', '.', '8', '.', '.', '7', '9']])

    runTestCase([["8","3",".",".","7",".",".",".","."],
                 ["6",".",".","1","9","5",".",".","."],
                 [".","9","8",".",".",".",".","6","."],
                 ["8",".",".",".","6",".",".",".","3"],
                 ["4",".",".","8",".","3",".",".","1"],
                 ["7",".",".",".","2",".",".",".","6"],
                 [".","6",".",".",".",".","2","8","."],
                 [".",".",".","4","1","9",".",".","5"],
                 [".",".",".",".","8",".",".","7","9"]])

